torch/csrc/generic/StorageSharing.cpp

```diff
@@ -212,27 +212,49 @@ static PyObject * THPStorage_(newSharedFd)(PyObject *_unused, PyObject *args)
 
 #else // THC_GENERIC_FILE
 
-static PyObject * THPStorage_(shareCuda)(THPStorage *self)
-{
+static PyObject* THPStorage_(shareCuda)(THPStorage* self) {
   HANDLE_TH_ERRORS
-  THWStorage *storage = self->cdata;
+  THWStorage* storage = self->cdata;
+
+  if (storage->received_cuda()) {
+    AT_ERROR(
+        "Attempting to send tensor received from another process. Consider clonning.");
+  }
+
   at::DeviceGuard device_guard(storage->device());
-  THPObjectPtr tuple(PyTuple_New(4));
+  THPObjectPtr tuple(PyTuple_New(5));
   THPObjectPtr device(PyLong_FromLong(storage->device().index()));
   THPObjectPtr _handle(Py_None);
   Py_INCREF(Py_None);
   THPObjectPtr size_bytes(PyLong_FromLong(storage->numel() * sizeof(scalar_t)));
   THPObjectPtr _offset_bytes(PyLong_FromLong(0));
+  THPObjectPtr _ref_counter(Py_None);
+  Py_INCREF(Py_None);
   if (THWStorage_(data)(LIBRARY_STATE storage)) {
     size_t base_size;
-    void *base_ptr = c10::cuda::CUDACachingAllocator::getBaseAllocation(THWStorage_(data)(LIBRARY_STATE storage), &base_size);
+    void* base_ptr = c10::cuda::CUDACachingAllocator::getBaseAllocation(
+        THWStorage_(data)(LIBRARY_STATE storage), &base_size);
     ptrdiff_t offset_bytes = (char*)storage->data<scalar_t>() - (char*)base_ptr;
 
     cudaIpcMemHandle_t handle;
     THCudaCheck(cudaIpcGetMemHandle(&handle, base_ptr));
 
-    _handle = PyBytes_FromStringAndSize((char *)&handle, CUDA_IPC_HANDLE_SIZE);
+    _handle = PyBytes_FromStringAndSize((char*)&handle, CUDA_IPC_HANDLE_SIZE);
     _offset_bytes = PyLong_FromSsize_t((Py_ssize_t)offset_bytes);
+
+    int flags = TH_ALLOCATOR_MAPPED_SHAREDMEM | TH_ALLOCATOR_MAPPED_EXCLUSIVE;
+    std::string ref_counter_handle = THPStorage_(__newHandle)();
+    at::DataPtr sptr = THRefcountedMapAllocator::makeDataPtr(
+        ref_counter_handle.c_str(), flags, sizeof(int64_t), nullptr);
```

Comment:
Re docs, for example, a little note here saying what exactly we are storing in `sptr` helps readers a lot! (And actually, I am personally a bit confused here: are we storing a reference count? Doesn't `THRefcountedMapAllocator` already handle that?)

Reply:
THRefcountedMapAllocator keeps number of open file/mem references. I agree it doesn't makes much sense now, but separation needed to support shared_mem/offset solution.